File: week4/Assignment/Question1.py
import sys
sys.stdin = open("input.txt", 'r')
sys.stdout = open("output.txt", 'w')
MOD = 10**9 + 7


class DSU:
    def __init__(self, n):
        self.parent = [i for i in range(n+1)]
        self.rank = [0]*(n+1)
        self.number_of_sets = n
        self.setsize = [1]*(n+1)

    def findset(self, u):
        if self.parent[u] == u:
            return u
        else:
            self.parent[u] = self.findset(self.parent[u])
            return self.parent[u]

    def unions(self, u, v):
        u, v = self.findset(u), self.findset(v)
        if u != v:
            if self.rank[u] > self.rank[v]:
                u, v = v, u
            if self.rank[u] == self.rank[v]:
                self.rank[v] += 1
            self.parent[u] = v
            self.number_of_sets -= 1
            self.setsize[v] += self.setsize[u]

    def total_sets(self):
        return self.number_of_sets

    def setsize(self, i):
        return self.setsize[self.findset(i)]


# Read the number of test cases
t = int(input())

# Process each test case
for _ in range(t):
    # Read the input for this test case
    n = int(input())
    A = list(map(int, input().split()))
    B = list(map(int, input().split()))
    X = list(map(int, input().split()))

    # Count the number of possible permutations of C
    ans = DSU(n)
    for i in range(n):
        ans.unions(A[i], B[i])
    unique = []
    for i in range(n):
        if ans.findset(i+1) not in unique and ans.setsize[ans.findset(i+1)] > 1:
            unique.append(ans.findset(i+1))

    # Output the result modulo 10^9+7
    print((2**len(unique)) % MOD)

File: week4/Assignment/Question1.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

const ll MOD = 1e9 + 7;

class DSU
{
public:
    vector<ll> parent, rank, setsize;
    ll number_of_sets;

    DSU(ll n)
    {
        parent.resize(n + 1);
        rank.resize(n + 1);
        setsize.resize(n + 1);
        for (ll i = 1; i <= n; i++)
        {
            parent[i] = i;
            rank[i] = 0;
            setsize[i] = 1;
        }
        number_of_sets = n;
    }

    ll findset(ll u)
    {
        if (parent[u] == u)
        {
            return u;
        }
        else
        {
            parent[u] = findset(parent[u]);
            return parent[u];
        }
    }

    void unions(ll u, ll v)
    {
        u = findset(u);
        v = findset(v);
        if (u != v)
        {
            if (rank[u] > rank[v])
            {
                swap(u, v);
            }
            if (rank[u] == rank[v])
            {
                rank[v]++;
            }
            parent[u] = v;
            number_of_sets--;
            setsize[v] += setsize[u];
        }
    }

    ll total_sets()
    {
        return number_of_sets;
    }

    ll get_setsize(ll i)
    {
        return setsize[findset(i)];
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);

    // Read the number of test cases
    ll t;
    cin >> t;

    // Process each test case
    while (t--)
    {
        // Read the input for this test case
        ll n;
        cin >> n;
        vector<ll> A(n), B(n), X(n);
        for (ll i = 0; i < n; i++)
        {
            cin >> A[i];
        }
        for (ll i = 0; i < n; i++)
        {
            cin >> B[i];
        }
        for (ll i = 0; i < n; i++)
        {
            cin >> X[i];
        }

        // Count the number of possible permutations of C
        DSU ans(n);
        for (ll i = 0; i < n; i++)
        {
            ans.unions(A[i], B[i]);
        }
        set<ll> unique;
        for (ll i = 1; i <= n; i++)
        {
            if (ans.get_setsize(i) > 1)
            {
                unique.insert(ans.findset(i));
            }
        }
        for (ll i = 0; i < n; i++)
        {
            if (X[i] != 0)
            {
                unique.erase(ans.findset(X[i]));
            }
        }
        // cout << (ll)(pow(2, unique.size())) % MOD << "\n";
        ll answer=1;
        for(ll i = 0; i < unique.size(); i++)
        {
            answer = (answer * 2) % MOD;
        }
        cout << answer << "\n";
    }

    return 0;
}
